### PR TITLE
WIP: Add breakpad (symbol dumping) support

### PR DESCRIPTION
**Pull request by dschmidt**

### CODE REVIEW

File: CraftSettings.ini.template [hunk 1]
```diff
@@ -230,6 +230,9 @@ Locality = Berlin
 Country = DE
 State =
 
+# Use dump_syms from breakpad-tools to dump debug symbols at post-install step
+[SymbolDumping]
+Enabled = False
```
Comment (TheOneRing) [inline CraftSettings.ini.template]: Just make it a setting in the Packager category

File: bin/Packager/CollectionPackagerBase.py [hunk 2]
```diff
@@ -256,6 +262,50 @@ def copyFiles(self, srcDir, destDir, dontStrip) -> bool:
...
+            firstLine = str(outBytes.splitlines()[0], 'utf-8')
+            CraftCore.log.debug('Module line: %s' % firstLine)
+            regex = "^MODULE [^ ]+ [^ ]+ ([0-9aA-fF]+) (.*)"
+            CraftCore.log.debug('regex: %s' % regex)
+            moduleLine = re.match(regex, firstLine)
+            CraftCore.log.debug('regex: %s' % moduleLine)
+            outputPath = os.path.join(self.symbolsDir(), moduleLine.group(2),
+                                 moduleLine.group(1))
+
+            if (not os.path.exists(outputPath)):
```
Comment (TheOneRing) [inline bin/Packager/CollectionPackagerBase.py]: just use utils.createDirectory without a condition

File: blueprints/dev-utils/breakpad-tools/breakpad-tools.py [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+        for ver in ["pecoff-dwarf-on-git-20171117-fetch-externals-on-win"]:
+            self.svnTargets[ver] = f"[git]https://github.com/dschmidt/google-breakpad|{ver}|"
+            self.patchToApply[ver] = [("0001-Add-CMake-script-to-build-dump_syms.patch", 1)]
+
+        self.defaultTarget = 'pecoff-dwarf-on-git-20171117-fetch-externals-on-win'
+        self.description = "The tools part of the breakpad crash-reporting system."
+        self.webpage = "https://github.com/dschmidt/google-breakpad"
+
+    def setDependencies(self):
+        self.buildDependencies["dev-utils/python2"] = "default"
```
Comment (TheOneRing) [inline blueprints/dev-utils/breakpad-tools/breakpad-tools.py]: also depend on virtual/base and use None instead of default

File: bin/info.py [hunk 4]
```diff
@@ -71,6 +71,11 @@ def __init__(self, parent):
         # do this after buildTarget is set so that some dependencies can be set depending on self.buildTarget
         self.setDependencies()
 
+        # Where to put this? How to make sure it's not required before it could be built?
+        # Should we set this before setDependencies() and disable it in all of its dependencies? :-\
+        if CraftCore.settings.getboolean("SymbolDumping", "Enabled", True) and not "dev-utils/breakpad-tools" in self.buildDependencies: # TODO: default to False
+            self.buildDependencies["dev-utils/breakpad-tools"] = "default"
```
Comment (TheOneRing) [inline bin/info.py]: None instead of default

File: bin/Packager/CollectionPackagerBase.py [hunk 5]
```diff
@@ -256,6 +262,50 @@ def copyFiles(self, srcDir, destDir, dontStrip) -> bool:
                     utils.sign([entry_target])
         return True
 
+    def symbolsDir(self):
+        """return absolute path to the symbol directory in imageDir
+        """
+        return os.path.join(self.imageDir(), 'symbols')
+
+    def internalPostInstall(self):
+        if not super().internalPostInstall():
```
Comment (TheOneRing) [inline bin/Packager/CollectionPackagerBase.py]: We need to move this to a class in a higher level. Not everyone is inheriting CollectionPackagerBase.
Mybe just use PackagerBase?

File: bin/Packager/CollectionPackagerBase.py [hunk 6]
```diff
@@ -256,6 +262,50 @@ def copyFiles(self, srcDir, destDir, dontStrip) -> bool:
...
+            return False
+
+        if self._doDumpSymbols:
+            for entry in self.traverse(self.imageDir(), self.whitelisted, self.blacklisted):
```
Comment (TheOneRing) [inline bin/Packager/CollectionPackagerBase.py]: We probably won't have a black or white list at the point we execute this code.
Please use os.walk for now or something similar

**Review by TheOneRing — CHANGES_REQUESTED** [review]

**Comment by TheOneRing** [comment]
Solved in https://github.com/owncloud/craft-blueprints-owncloud

**Comment by dschmidt** [comment]
It isn't but I won't fight over this as Sentry has native .pdb and .dSYM support.

FTR; This PR was about adding breakpad symbol dumping to all packages, so we don't keep dumping Qt symbols (e.g.) over and over again when we actually just would need to dump the new build client symbols. That's why I wanted to bring this to the core ...

**Pull request closed without merging**